env/env_inspected.cc

```diff
@@ -0,0 +1,341 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+#ifndef ROCKSDB_LITE
+#include "rocksdb/env_inspected.h"
+
+namespace rocksdb {
+
+class InspectedSequentialFile : public SequentialFileWrapper {
+ public:
+  InspectedSequentialFile(std::unique_ptr<SequentialFile>&& target,
+                          FileSystemInspector* inspector)
+      : SequentialFileWrapper(target.get()),
+        owner_(std::move(target)),
+        inspector_(inspector) {}
+
+  Status Read(size_t n, Slice* result, char* scratch) override {
+    assert(inspector_);
+    Status s;
+    size_t offset = 0;
+    size_t allowed = 0;
+    while (offset < n) {
+      allowed = inspector_->Read(Env::IO_UNCATEGORIZED, n - offset);
+      if (allowed > 0) {
+        s = SequentialFileWrapper::Read(allowed, result, scratch + offset);
+        if (!s.ok()) {
+          break;
+        }
+        size_t actual_read = result->size();
+        if (result->data() != scratch + offset) {
```

Comment:
when can this happen?

Reply:
It's possible for underlying implementation to store the data in their own private buffer, instead of using user provided `scratch`.